Transactions

reaperzn opened the pull request:
Proposed code for  #21

Commit:
Added an equivalent of transactions back to auter

Commit:
Adjusted the position of the definition of TRANSACTIONCOMMAND

Commit:
Re-ordered some code in the apply section

piersc commented:
- Probably we should set /var/cache/auter/$CONFIGSET as a constant, and manipulate PACKAGEMANAGEROPTIONS options outside the functions
- TRANSACTIONCOMMAND this is really "downloaddir" so name it like that
- We should probably check mode and ownership if this dir exists for safety
- Line 149: its not an error if this dir doesn't exist (if on first run no updates are available, this won't get created during --prep). I don't think we care if it doesn't exist, just means there's nothing to do.
- Remove the extra line feed on line 332
- INFO: Updates downloaded should say where its been downloaded to if not default
- spec file will need to be updated, and own the cache dir

Commit:
Addressed comments made by piersc in #22

1) /var/cache/auter/ has been set as a variable called DOWNLOADDIR
2) TRANSACTIONCOMMAND has now been added to PACKAGEMANAGEROPTIONS in line 326, after CONFIGSET has been initialised
3) Added a check for directory permissions
4) The error message has been removed
5) Extra line feed has been removed
6) Added the directory to the INFO message

Commit:
Updated spec file to include /var/cache/auter

piersc commented:
Looks like your last commit overwrote some of your prior changes

Commit:
Renamed USETRANSACTIONS to ONLYINSTALLFROMPREP

reaperzn commented:
I Renamed USETRANSACTIONS to ONLYINSTALLFROMPREP and also used a new RPMS variable which more accurately describes the contents. 

I also previously forgot to add the variables to the update command (currently line 167)

piersc commented:
Typically, you wouldn't include a trailing slash on a path (DOWNLOADDIR)
  rm -f ${DOWNLOADDIR}/${CONFIGSET}/_.rpm
    becomes
  rm -f /var/cache/auter//default/_.rpm

To avoid writing whitespace at the end of the log msg, change:
  logit "INFO: Updates downloaded ${DOWNLOADLOGMSG}"
to:
  logit "INFO: Updates downloaded${DOWNLOADLOGMSG}"

Remove the extra space on line 142

Log msg on line 147 needs to be shorter (<= 70 chars) so it doesn't wrap when viewing logs

No real need to instantiate $RPMs on line 137 (if you're not doing it for $DOWNLOADLOGMSG)

Why not do the directory create/check mode in the main section where PACKAGEMANAGEROPTIONS is set

Commit:
Addressed latest comment from piersc:
1) Removed the trailing / from $DOWNLOADDIR
2) Adjusted Info message to remove whitespace
3) Removed extra line space in line 142
4) Shortened error message on line 147
5) removed instantiation of $RPMS
6) Moved the directory check

Commit:
DNF support for INSTALLONLYONPREP

piersc commented:
Changes looked good for yum, but it seems unfortunately dnf doesn't support the --downloaddir option: 
https://bugzilla.redhat.com/show_bug.cgi?id=1203491

I've updated your PR with a first stab at making this work for dnf:
- If ONLYINSTALLFROMPREP and DNF, then mv rpms from /var/cache/dnf to the right place
- Change the verb to pass to pkgmanager from update to install for RPMs, otherwise you get errors like this:

```
Package mpfr not installed, cannot update it.
```

Tested on Fedora 24 only, need to test on CentOS 6 and 7 as well to ensure this didn't break anything.

Beerey commented on auter:
```diff
@@ -106,10 +107,22 @@ function prepare_updates() {
-        PREPOUTPUT=$(${PACKAGE_MANAGER} ${PACKAGEMANAGEROPTIONS} update --downloadonly -y)
-        logit "INFO: Updates downloaded"
+        if [[ "${ONLYINSTALLFROMPREP}" == "yes" ]]; then
+          rm -f ${DOWNLOADDIR}/${CONFIGSET}/*.rpm
```
Quote these variables.

reaperzn replied:
fixed in next commit

Beerey commented on auter:
```diff
@@ -106,10 +107,22 @@ function prepare_updates() {
+            find /var/cache/dnf -name *.rpm -exec rm -f {} \;
+            DOWNLOADOPTION=""
+          fi
+          DOWNLOADLOGMSG=" to ${DOWNLOADDIR}/${CONFIGSET}"
```
Pedantism, but it seems weird to have "to " here.  Why not add it to the actual logit message?

reaperzn replied:
This variable is used later on (line 123). If this variable is set then the logit message reads:
"INFO: Updates downloaded to /var/cache/auter/default"

If it is not set (ie when $ONLYINSTALLFROMPREP == no) it will read:
"INFO: Updates downloaded"

If I take out the " to" then we will end up with:
"INFO: Updates downloaded /var/cache/auter/default"

I am open to suggestions...

Beerey commented on auter:
```diff
@@ -301,6 +323,21 @@ if [[ -z "${CONFIGSET}" ]]; then
+    logit "ERROR: ${DOWNLOADDIR}/${CONFIGSET} does not have the correct permissions."
+    quit 3
+  fi
+  DOWNLOADOPTION="--downloaddir=${DOWNLOADDIR}/${CONFIGSET}"
```
I don't like setting DOWNLOADOPTION here, when we're actually only using the variable within the prep_updates function.  Could we move this to the logic where it's actually being used?

Even better would be to just use the download to cache and move to the final location method for both yum and dnf - reduces code complexity, even if it's not quite as elegant as --downloaddir.

Same story for setting the UPDATEACTION variable.  Logic related to what we do to prep/apply updates should generally be in the relevant function instead of main().

reaperzn replied:
DOWNLOADOPTION has been moved to prepare_updates function.
UPDATEACTION check has now been moved to the apply_updates function.

We should be using built-in functionality of yum. The fact that dnf does not yet have this feature should not make us disregard built in functions of yum.

Beerey commented on auter.conf:
```diff
@@ -19,6 +19,14 @@ PACKAGEMANAGEROPTIONS=""
 # during --prep
 PREDOWNLOADUPDATES="yes"
 
+# Option to only install the downloaded packages. This option is useful if you
```
I think we could be more explicit about the use case here?  ie "This option should be used when getting multiple servers to apply the same update list (eg to keep a uat/staging/prod environment in sync), or when you would like to review the updates ahead of applying them.  Not setting this option will result in auter installing the latest available packages at the time apply runs."  Maybe something along those lines?

reaperzn replied:
Updated

Beerey commented on auter.help2man:
```diff
@@ -27,6 +27,9 @@
   Directories containing scripts to run before/after updates are applied, and before/after the server is rebooted (if applicable). Default locations are above, but also customizable in the config file. If any scripts in the pre/post hook directories are not executable a warning will be logged with the filename but auter will continue. If any scripts in the pre-post hook directories exit with a no ...
 
+\//var/cache/auter/
+  Will be created during the prep stage if ONLYINSTALLFROMPREP is set to 'yes'. A configset sub-directory will be created which will hold the updates that were downloaded from the most recent --prep execution. Any existing rpm files will be deleted as part of the following --prep execution. The directories for each profile must have owner and group set to root and must NOT be writable by other.
```
This is technically not true according to the spec file - the directory will be created at install time.

reaperzn replied:
Updated

Commit:
Addressed comments by Beerey in #22

reaperzn commented:
@Beerey Please could you review the latest updates in this PR

Beerey commented:
Ok to merge from me.
